File: assignment1/r/part5.R
# ----------------------------------------------------------------
# A. 修正后的数据加载
# ----------------------------------------------------------------

# 1. 设置 header=TRUE 以将第一行识别为列名并跳过。
# 2. [, 1] 选取第一列。
# 3. as.numeric() 显式确保数据是数值类型。
data <- as.numeric(read.csv("DataPart5.csv", header=TRUE)[, 1]) 

# ----------------------------------------------------------------
# B. 拟合 ARMA(2,1) 模型
# ----------------------------------------------------------------

# 使用 arima() 函数拟合 ARMA(2,1) 模型
model_arma21 <- arima(data, order = c(2, 0, 1))

print("--- ARMA(2,1) Model Fit Summary ---")
print(model_arma21)

# 提取模型残差
residuals_arma21 <- residuals(model_arma21)

# Ljung-Box 检验 (可选但推荐)
Box.test(residuals_arma21, lag = 20, type = "Ljung-Box")


# ----------------------------------------------------------------
# C. 对残差运行 LDF
# ----------------------------------------------------------------

# x: 模型的残差 (e_t)
# lags: 想要检验的滞后数 (例如，检验前 10 个滞后)
# nBoot: 用于计算置信带的 Bootstrap 次数 (建议 100 或更高)

print("--- Running LDF on ARMA(2,1) Residuals ---")

ldf_results <- ldf(
  x = residuals_arma21, 
  lags = 10, 
  nBoot = 100, 
  plotIt = TRUE, # 绘制 LDF 图
  plotFits = FALSE
)

# ----------------------------------------------------------------
# D. 分析与改进模型
# ----------------------------------------------------------------

# 1. 查看生成的 LDF 图。如果 LDF 曲线在任何滞后 n 处
#    显著超出置信带，则表示残差中存在显著的非线性。
# 2. 找到显著的滞后 n。
# 3. 绘制诊断图 e_t vs. e_{t-n}：
#    plot(residuals_arma21[-c(1:n)], residuals_arma21[-c((n+1):length(residuals_arma21))],
#         xlab=paste0("e_{t-", n, "}"), ylab="e_t", main=paste0("Residuals: e_t vs. e_{t-", n, "}"))
# 4. 根据诊断图的形状（例如，SETAR 的 V 形、ARCH 的喇叭形），
#    提出一个更合适的非线性模型结构（如 SETAR, ARCH/GARCH 等）。

File: assignment1/Part 5.py
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.graphics.tsaplots import plot_acf as sm_plot_acf, plot_pacf as sm_plot_pacf
from statsmodels.graphics.gofplots import qqplot  # <— for QQ plot
from LDF import ldf


data = pd.read_csv("DataPart5.csv")
y = data.iloc[:, 0].astype(float).dropna()

# Model orders (increasing complexity)
orders = [(1,1), (2,1), (2,2)]  # (p,q)

models, residuals = [], []
for p, q in orders:
    m = ARIMA(y, order=(p, 0, q)).fit()
    models.append(m)
    residuals.append(m.resid)

fig, axs = plt.subplots(nrows=3, ncols=4, figsize=(17, 10), constrained_layout=True)

for i, ((p, q), resid) in enumerate(zip(orders, residuals)):

    sm_plot_acf(resid, lags=30, ax=axs[i, 0])
    axs[i, 0].set_title(f"ACF — ARMA({p},{q}) residuals")
    axs[2, 0].set_xlabel("lag")

    sm_plot_pacf(resid, lags=30, ax=axs[i, 1], method="ywm")
    axs[i, 1].set_title(f"PACF — ARMA({p},{q}) residuals")
    axs[2, 1].set_xlabel("lag")

    axs[i, 2].plot(resid, lw=1)
    axs[i, 2].axhline(0, ls="--", lw=1, color="gray")
    axs[i, 2].set_title(f"Residuals — ARMA({p},{q})")
    axs[2, 2].set_xlabel("t")
    axs[i, 2].set_ylabel("Residual")

 
    qqplot(resid, line="45", ax=axs[i, 3], color="C1", alpha=0.8)
    axs[i, 3].set_title(f"QQ plot — ARMA({p},{q})")

# Shared labels/titles
fig.suptitle("ARMA model diagnostics by order : ACF, PACF, Residuals, QQ-plot", fontsize=13)
plt.tight_layout(pad=2.0)
plt.savefig("Part 5 ARMA diagnostics.png", dpi=300)


x = residuals[1]  # index 1 corresponds to ARMA(2,1)
lags = [1, 2, 3, 4, 5, 10, 20]

vals, q95 = ldf(x, lags, nBoot=20, plotIt=True, random_state=42)
plt.title("LDF of ARMA(2,1) residuals")
plt.savefig("Part 5 LDF residuals.png", dpi=300)
plt.show()


